<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Ready-to-Use Omics Formatting, Analysis, and Visualization Pipeline • readyomics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Ready-to-Use Omics Formatting, Analysis, and Visualization Pipeline">
<meta name="description" content="Provides a flexible and streamlined pipeline for formatting, analyzing, and visualizing omics data, regardless of omics type (e.g. transcriptomics, proteomics, metabolomics). The package includes tools for shaping input data into analysis-ready structures, fitting linear or mixed-effect models, extracting key contrasts, and generating a rich variety of ready-to-use publication-quality plots. Designed for transparency and reproducibility across a wide range of study designs, with customizable components for statistical modeling.">
<meta property="og:description" content="Provides a flexible and streamlined pipeline for formatting, analyzing, and visualizing omics data, regardless of omics type (e.g. transcriptomics, proteomics, metabolomics). The package includes tools for shaping input data into analysis-ready structures, fitting linear or mixed-effect models, extracting key contrasts, and generating a rich variety of ready-to-use publication-quality plots. Designed for transparency and reproducibility across a wide range of study designs, with customizable components for statistical modeling.">
<meta property="og:image" content="https://lmartinezgili.github.io/readyomics/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">readyomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/readyomics.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/lmartinezgili/readyomics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="readyomics-">readyomics <a class="anchor" aria-label="anchor" href="#readyomics-"></a>
</h1>
</div>
<!-- badges: start -->

<p>readyomics provides a pipeline for formatting, analyzing, and visualizing omics data - regardless of omics type (e.g. transcriptomics, proteomics, metabolomics, metagenomics).</p>
<p>It is designed for flexibility, reproducibility, and scalability across a wide range of study designs, with modular components for statistical analysis and visualization.</p>
<p>It includes tools to:</p>
<ul>
<li>Process data into analysis-ready.</li>
<li>Perform multivariate analysis.</li>
<li>Fit linear or mixed-effects models.</li>
<li>Produce publication-quality plots.</li>
</ul>
<div class="section level2">
<h2 id="essential-standard-terms-used-in-the-package">Essential standard terms used in the package<a class="anchor" aria-label="anchor" href="#essential-standard-terms-used-in-the-package"></a>
</h2>
<ul>
<li>
<strong>Platform</strong>: the technology or instrument used to generate omics data, such as next-generation sequencing (NGS), mass spectrometry (MS), or nuclear magnetic resonance spectroscopy (NMR).</li>
<li>
<strong>Feature(s)</strong>: a general term for a biological variable that has been profiled with an omics platform, such as metabolites, lipids, genes, proteins, or microbial taxa, depending on the assay.</li>
<li>
<strong>Sample data</strong> (!= <em>metadata</em>*): biological or demographic information collected for each study sample (e.g., experimental group, age, sex, BMI).</li>
</ul>
<p>*<em>Note</em>: in its strict sense, metadata (“data about data”) refers to information describing the context, structure, or properties of a dataset — such as acquisition date, instrument settings, plate/well ID, or run order. It does not refer to biological or demographic variables. To avoid ambiguity, readyomics adopts the same convention as phyloseq, using the term <em>sample data</em> for variables describing the study samples.</p>
</div>
<div class="section level2">
<h2 id="main-functionalities">Main functionalities<a class="anchor" aria-label="anchor" href="#main-functionalities"></a>
</h2>
<div class="section level3">
<h3 id="data-processing-normalisation-transformation-filtering">Data processing (normalisation, transformation, filtering)<a class="anchor" aria-label="anchor" href="#data-processing-normalisation-transformation-filtering"></a>
</h3>
<ul>
<li>
<code><a href="reference/process_ngs.html">process_ngs()</a></code>: process next-generation sequencing data.</li>
<li>
<code><a href="reference/process_ms.html">process_ms()</a></code>: process MS or NMR data.</li>
<li>
<code><a href="reference/build_phyloseq.html">build_phyloseq()</a></code>: build phyloseq objects for metataxonomic data.</li>
</ul>
</div>
<div class="section level3">
<h3 id="multivariate-analysis">Multivariate analysis<a class="anchor" aria-label="anchor" href="#multivariate-analysis"></a>
</h3>
<ul>
<li>
<code><a href="reference/mva.html">mva()</a></code>: PCA, PLS and OPLS models.</li>
<li>
<code><a href="reference/permanova.html">permanova()</a></code>: wrapper for <code><a href="https://vegandevs.github.io/vegan/reference/adonis.html" class="external-link">vegan::adonis2()</a></code> function with additional options and summary results.</li>
</ul>
</div>
<div class="section level3">
<h3 id="differential-abundanceexpression-analysis">Differential [abundance/expression] analysis<a class="anchor" aria-label="anchor" href="#differential-abundanceexpression-analysis"></a>
</h3>
<ul>
<li>
<code><a href="reference/dana.html">dana()</a></code>: fit feature-wise linear fixed or mixed effects models.</li>
<li>
<code><a href="reference/adjust_pval.html">adjust_pval()</a></code>: methods to adjust nominal P-values on <code><a href="reference/dana.html">dana()</a></code> result.</li>
<li>
<code><a href="reference/ready_plots.html">ready_plots()</a></code>: visualize <code>dana</code> results and significant features.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"readyomics"</span><span class="op">)</span></span></code></pre></div>
<p>You can install the development version of readyomics from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"lmartinezgili/readyomics"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-started">Get started<a class="anchor" aria-label="anchor" href="#get-started"></a>
</h2>
<div class="section level3">
<h3 id="types-of-input-data">Types of input data<a class="anchor" aria-label="anchor" href="#types-of-input-data"></a>
</h3>
<p>readyomics is as omics-agnostic and inclusive as possible.</p>
<p>Raw omics files (e.g., .fastq, .mzML) must first be pre-processed through external commercial or open-source pipelines into a data matrix where each row corresponds to a sample and each column corresponds to a measured omics feature.</p>
</div>
<div class="section level3">
<h3 id="required-files-and-format">Required files and format<a class="anchor" aria-label="anchor" href="#required-files-and-format"></a>
</h3>
<ul>
<li>
<code>X</code>: a .csv or .RDS table of omics data (samples in rows and features in columns).</li>
<li>
<code>sample_data</code> (or <code>sdata</code>): a .csv or .RDS table of study sample information (samples in rows).</li>
<li>
<code>sample_id</code> must be a column in <code>sample_data</code> and have unique ids for each sample.</li>
<li>Row names in <code>X</code> and <code>sample_data</code> must match <code>sample_id</code> values, though order can differ — readyomics functions will check and align automatically.</li>
</ul>
</div>
<div class="section level3">
<h3 id="documentation-and-examples">Documentation and Examples<a class="anchor" aria-label="anchor" href="#documentation-and-examples"></a>
</h3>
<p>For tutorials, examples, and reference documentation, visit readyomics <a href="https://lmartinezgili.github.io/readyomics/">website</a>.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=readyomics" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/lmartinezgili/readyomics/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/lmartinezgili/readyomics/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing readyomics</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Laura Martinez-Gili <br><small class="roles"> Maintainer, author, copyright holder </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/lmartinezgili/readyomics/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/lmartinezgili/readyomics/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Laura Martinez-Gili.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
