<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>readyomics • readyomics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="readyomics">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">readyomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/readyomics.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/lmartinezgili/readyomics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>readyomics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/lmartinezgili/readyomics/blob/v0.1.1/vignettes/readyomics.Rmd" class="external-link"><code>vignettes/readyomics.Rmd</code></a></small>
      <div class="d-none name"><code>readyomics.Rmd</code></div>
    </div>

    
    
<p>readyomics streamlines single- or multi-omics data analysis
pipelines.</p>
<p>The starting input should be a pre-processed raw omics data matrix
(e.g. gene counts, metabolites or protein abundances, etc.) and a table
of related sample information to be used for additional processing and
statistical analysis.</p>
<p>readyomics splits processing and statistical analysis steps into
different functions. This has the following advantages:</p>
<ul>
<li>
<strong>Aligns with best practice in data science</strong>
<ul>
<li>Avoiding “all-in-one” black-box functions.</li>
<li>Processing functions (<code><a href="../reference/process_ngs.html">process_ngs()</a></code>,
<code><a href="../reference/process_ms.html">process_ms()</a></code>) prepare data.</li>
<li>Analysis and visualization functions (<code><a href="../reference/dana.html">dana()</a></code>,
<code><a href="../reference/permanova.html">permanova()</a></code> and <code><a href="../reference/ready_plots.html">ready_plots()</a></code>) use data ready
for modeling.</li>
</ul>
</li>
<li>
<strong>Transparency and troubleshooting</strong>
<ul>
<li>Makes it easy to pinpoint where things go wrong (e.g.,
transformation <em>vs.</em> model error).</li>
<li>Encourages users to inspect data at each stage.</li>
</ul>
</li>
<li>
<strong>Supports nonlinear or custom pipelines</strong>
<ul>
<li>Externally-formatted data ⟶ statistical analysis with
readyomics.</li>
<li>Process raw data with readyomics ⟶ use other downstream analysis
methods.</li>
</ul>
</li>
<li>
<strong>Scalability</strong>
<ul>
<li>
<code><a href="../reference/dana.html">dana()</a></code> can be set up to run in parallel through
<code><a href="https://future.futureverse.org/reference/plan.html" class="external-link">future::plan()</a></code>.</li>
<li>
<code>dana</code> results from multiple omics can be joined, using
<code>platform</code> and <code>assay</code> to distinguish results from
each dataset.</li>
<li>Effortless <code>dana</code> results analysis by piping
<code><a href="../reference/adjust_pval.html">adjust_pval()</a></code>, <code><a href="../reference/ready_plots.html">ready_plots()</a></code>, and (optionally)
also <code><a href="../reference/add_feat_name.html">add_feat_name()</a></code> or <code><a href="../reference/add_taxa.html">add_taxa()</a></code>.</li>
</ul>
</li>
<li>
<strong>Future package extensions</strong>
<ul>
<li>New pre-processing strategies.</li>
<li>New analysis methods.</li>
</ul>
</li>
</ul>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>The following example shows the overall workflow of readyomics.</p>
<p>The example data is a 16S rRNA gene sequencing metataxonomics dataset
from human fecal samples. They are a subset of 29 samples from a T2DM
prospective cohort screened for MASLD:</p>
<blockquote>
<p>Forlano, R.*, Martinez-Gili, L.*, Takis, P., Miguens-Blanco, J., Liu,
T., Triantafyllou, E., … Manousou, P. (2024). Disruption of gut barrier
integrity and host–microbiome interactions underlie MASLD severity in
patients with type-2 diabetes mellitus. Gut Microbes, 16(1). <a href="https://doi.org/10.1080/19490976.2024.2304157" class="external-link uri">https://doi.org/10.1080/19490976.2024.2304157</a></p>
</blockquote>
<p>Example input files can be found in the <code>inst/extdata</code>
folder of the readyomics source, and can be imported using
<code><a href="https://rdrr.io/r/base/system.file.html" class="external-link">base::system.file()</a></code> command as shown below in <a href="#import-and-inspect-data">Import and inspect data</a>.</p>
<p><strong>Note</strong>: this example is meant to showcase basic
package usage. It is not meant to provide guidelines on suitable
study-specific choices for data processing and analysis.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"readyomics"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-the-package">Load the package<a class="anchor" aria-label="anchor" href="#load-the-package"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lmartinezgili.github.io/readyomics/">readyomics</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="import-and-inspect-data">Import and inspect data<a class="anchor" aria-label="anchor" href="#import-and-inspect-data"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Raw matrix of ASV counts</span></span>
<span><span class="va">asv_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"asv_raw_counts.csv"</span>, package <span class="op">=</span> <span class="st">"readyomics"</span><span class="op">)</span>, </span>
<span>                       check.names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                       row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Taxonomy table</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"taxonomy.csv"</span>, package <span class="op">=</span> <span class="st">"readyomics"</span><span class="op">)</span>, </span>
<span>                 check.names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                 row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Sample data</span></span>
<span><span class="va">sample_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sample_data.csv"</span>, package <span class="op">=</span> <span class="st">"readyomics"</span><span class="op">)</span>, </span>
<span>                        check.names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                        row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">asv_counts</span><span class="op">)</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table style="width:100%;" class="table">
<colgroup>
<col width="1%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">RF_029_AKC</th>
<th align="right">RF_055_RJ</th>
<th align="right">RF_061_AJ</th>
<th align="right">RF_071_IT</th>
<th align="right">RF_081_MS</th>
<th align="right">RF_083_SS</th>
<th align="right">RF_089_RT</th>
<th align="right">RF_091_KB</th>
<th align="right">RF_101_HJ</th>
<th align="right">RF_108_RA</th>
<th align="right">RF_110_SE</th>
<th align="right">RF_113_SV</th>
<th align="right">RF_114_NJ</th>
<th align="right">RF_141_AMM</th>
<th align="right">RF_154_SP</th>
<th align="right">RF_158_AK</th>
<th align="right">RF_159_KH</th>
<th align="right">RF_173_DP</th>
<th align="right">RF_175_MS</th>
<th align="right">RF_189_BN</th>
<th align="right">RF_200_KAP</th>
<th align="right">RF_210_RWP</th>
<th align="right">RF_212_FF</th>
<th align="right">RF_219_ME</th>
<th align="right">RF_231_ED</th>
<th align="right">RF_232_HB</th>
<th align="right">RF_233_AN</th>
<th align="right">RF_238_AF</th>
<th align="right">RF_240_EJB</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ASV1</td>
<td align="right">140</td>
<td align="right">288</td>
<td align="right">249</td>
<td align="right">201</td>
<td align="right">349</td>
<td align="right">289</td>
<td align="right">633</td>
<td align="right">386</td>
<td align="right">572</td>
<td align="right">498</td>
<td align="right">487</td>
<td align="right">270</td>
<td align="right">220</td>
<td align="right">265</td>
<td align="right">439</td>
<td align="right">395</td>
<td align="right">613</td>
<td align="right">433</td>
<td align="right">317</td>
<td align="right">182</td>
<td align="right">93</td>
<td align="right">13</td>
<td align="right">51</td>
<td align="right">50</td>
<td align="right">122</td>
<td align="right">264</td>
<td align="right">305</td>
<td align="right">296</td>
<td align="right">80</td>
</tr>
<tr class="even">
<td align="left">ASV2</td>
<td align="right">12</td>
<td align="right">13</td>
<td align="right">3654</td>
<td align="right">14</td>
<td align="right">28</td>
<td align="right">37</td>
<td align="right">56</td>
<td align="right">12</td>
<td align="right">33</td>
<td align="right">69</td>
<td align="right">73</td>
<td align="right">15</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">34</td>
<td align="right">37</td>
<td align="right">1058</td>
<td align="right">50</td>
<td align="right">46</td>
<td align="right">13</td>
<td align="right">773</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1270</td>
<td align="right">12</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1374</td>
<td align="right">711</td>
</tr>
<tr class="odd">
<td align="left">ASV3</td>
<td align="right">34</td>
<td align="right">390</td>
<td align="right">533</td>
<td align="right">297</td>
<td align="right">485</td>
<td align="right">520</td>
<td align="right">1080</td>
<td align="right">285</td>
<td align="right">850</td>
<td align="right">587</td>
<td align="right">766</td>
<td align="right">542</td>
<td align="right">345</td>
<td align="right">355</td>
<td align="right">411</td>
<td align="right">342</td>
<td align="right">548</td>
<td align="right">827</td>
<td align="right">294</td>
<td align="right">251</td>
<td align="right">194</td>
<td align="right">0</td>
<td align="right">56</td>
<td align="right">79</td>
<td align="right">139</td>
<td align="right">337</td>
<td align="right">215</td>
<td align="right">194</td>
<td align="right">103</td>
</tr>
<tr class="even">
<td align="left">ASV4</td>
<td align="right">235</td>
<td align="right">67</td>
<td align="right">176</td>
<td align="right">38</td>
<td align="right">164</td>
<td align="right">71</td>
<td align="right">150</td>
<td align="right">74</td>
<td align="right">117</td>
<td align="right">136</td>
<td align="right">96</td>
<td align="right">66</td>
<td align="right">52</td>
<td align="right">247</td>
<td align="right">55</td>
<td align="right">92</td>
<td align="right">225</td>
<td align="right">111</td>
<td align="right">112</td>
<td align="right">106</td>
<td align="right">729</td>
<td align="right">34</td>
<td align="right">685</td>
<td align="right">0</td>
<td align="right">60</td>
<td align="right">71</td>
<td align="right">440</td>
<td align="right">0</td>
<td align="right">543</td>
</tr>
<tr class="odd">
<td align="left">ASV5</td>
<td align="right">88</td>
<td align="right">142</td>
<td align="right">131</td>
<td align="right">99</td>
<td align="right">150</td>
<td align="right">150</td>
<td align="right">300</td>
<td align="right">197</td>
<td align="right">222</td>
<td align="right">262</td>
<td align="right">243</td>
<td align="right">124</td>
<td align="right">83</td>
<td align="right">144</td>
<td align="right">116</td>
<td align="right">182</td>
<td align="right">135</td>
<td align="right">243</td>
<td align="right">84</td>
<td align="right">59</td>
<td align="right">25</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">44</td>
<td align="right">119</td>
<td align="right">40</td>
<td align="right">86</td>
<td align="right">45</td>
</tr>
<tr class="even">
<td align="left">ASV6</td>
<td align="right">63</td>
<td align="right">160</td>
<td align="right">248</td>
<td align="right">135</td>
<td align="right">212</td>
<td align="right">272</td>
<td align="right">532</td>
<td align="right">171</td>
<td align="right">371</td>
<td align="right">364</td>
<td align="right">267</td>
<td align="right">218</td>
<td align="right">170</td>
<td align="right">228</td>
<td align="right">164</td>
<td align="right">182</td>
<td align="right">289</td>
<td align="right">392</td>
<td align="right">146</td>
<td align="right">131</td>
<td align="right">38</td>
<td align="right">0</td>
<td align="right">25</td>
<td align="right">31</td>
<td align="right">44</td>
<td align="right">138</td>
<td align="right">98</td>
<td align="right">118</td>
<td align="right">61</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table style="width:100%;" class="table">
<colgroup>
<col width="3%">
<col width="6%">
<col width="10%">
<col width="14%">
<col width="12%">
<col width="13%">
<col width="15%">
<col width="23%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">Kingdom</th>
<th align="left">Phylum</th>
<th align="left">Class</th>
<th align="left">Order</th>
<th align="left">Family</th>
<th align="left">Genus</th>
<th align="left">Species</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ASV1</td>
<td align="left">Bacteria</td>
<td align="left">Bacteroidota</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroides</td>
<td align="left">vulgatus</td>
</tr>
<tr class="even">
<td align="left">ASV2</td>
<td align="left">Bacteria</td>
<td align="left">Bacteroidota</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroides</td>
<td align="left">dorei</td>
</tr>
<tr class="odd">
<td align="left">ASV3</td>
<td align="left">Bacteria</td>
<td align="left">Proteobacteria</td>
<td align="left">Gammaproteobacteria</td>
<td align="left">Enterobacterales</td>
<td align="left">Enterobacteriaceae</td>
<td align="left">Escherichia/Shigella</td>
<td align="left">Escherichia/Shigella(g)_unclass.</td>
</tr>
<tr class="even">
<td align="left">ASV4</td>
<td align="left">Bacteria</td>
<td align="left">Bacteroidota</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroides</td>
<td align="left">uniformis</td>
</tr>
<tr class="odd">
<td align="left">ASV5</td>
<td align="left">Bacteria</td>
<td align="left">Bacteroidota</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroides</td>
<td align="left">vulgatus</td>
</tr>
<tr class="even">
<td align="left">ASV6</td>
<td align="left">Bacteria</td>
<td align="left">Firmicutes</td>
<td align="left">Bacilli</td>
<td align="left">Lactobacillales</td>
<td align="left">Streptococcaceae</td>
<td align="left">Streptococcus</td>
<td align="left">Streptococcus(g)_unclass.</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">groups_liver</th>
<th align="right">sex</th>
<th align="right">age</th>
<th align="right">PPI</th>
<th align="right">smoking</th>
<th align="left">alcohol_imp</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">RF_029_AKC</td>
<td align="left">FL</td>
<td align="right">1</td>
<td align="right">73</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">abstinent</td>
</tr>
<tr class="even">
<td align="left">RF_055_RJ</td>
<td align="left">FL_HS</td>
<td align="right">2</td>
<td align="right">55</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="left">abstinent</td>
</tr>
<tr class="odd">
<td align="left">RF_061_AJ</td>
<td align="left">FL</td>
<td align="right">1</td>
<td align="right">71</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">abstinent</td>
</tr>
<tr class="even">
<td align="left">RF_071_IT</td>
<td align="left">FL_HS</td>
<td align="right">1</td>
<td align="right">71</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="left">moderate</td>
</tr>
<tr class="odd">
<td align="left">RF_081_MS</td>
<td align="left">FL</td>
<td align="right">1</td>
<td align="right">73</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">excess</td>
</tr>
<tr class="even">
<td align="left">RF_083_SS</td>
<td align="left">FL</td>
<td align="right">1</td>
<td align="right">70</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="left">abstinent</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="format-sample_data">Format sample_data<a class="anchor" aria-label="anchor" href="#format-sample_data"></a>
</h3>
<p><code>sample_data</code> must include a column named
<strong><code>sample_id</code></strong> with unique IDs for each sample
in the study. For longitudinal studies, <code>sample_id</code> still
needs to be unique, therefore a column such as <code>subject_id</code>
can then be used to indicate samples coming from the same
individual.</p>
<p>The omics <code>X</code> data matrix must also use the same
<code>sample_id</code> labels as row names, so that they can be matched
before analysis. readyomics always checks sample ID matching.</p>
<p>In addition, categorical variables must be set as factors for
downstream statistical analyses.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_data</span> <span class="op">&lt;-</span> <span class="va">sample_data</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">)</span>,</span>
<span>         groups_liver <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">groups_liver</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"FL_HS"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>         sex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">sex</span><span class="op">)</span>,</span>
<span>         PPI <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">PPI</span><span class="op">)</span>,</span>
<span>         smoking <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">smoking</span><span class="op">)</span>,</span>
<span>         alcohol_imp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">alcohol_imp</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="process-data-using-process_ngs">Process data using <code>process_ngs()</code><a class="anchor" aria-label="anchor" href="#process-data-using-process_ngs"></a>
</h3>
<p>The choice of processing function will depend on the type of omics
being analysed. We will use <code><a href="../reference/process_ngs.html">process_ngs()</a></code> in the example,
as it is suitable for metataxonomics.</p>
<p>See also <code><a href="../reference/process_ms.html">process_ms()</a></code> if you need to process mass
spectrometry (MS) data like metabolomics or proteomics.
<code><a href="../reference/process_ms.html">process_ms()</a></code> is compatible with nuclear magnetic resonance
spectroscopy (NMR) data, although with limited processing options.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Samples as rows required for process_ngs</span></span>
<span><span class="va">asv_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">asv_counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Process asv_counts data</span></span>
<span><span class="va">asv_ready</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_ngs.html">process_ngs</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">asv_counts</span>, </span>
<span>                         sample_data <span class="op">=</span> <span class="va">sample_data</span>, </span>
<span>                         taxa_table <span class="op">=</span> <span class="va">taxa</span>, </span>
<span>                         normalise <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                         transform <span class="op">=</span> <span class="st">"clr"</span>, </span>
<span>                         eco_phyloseq <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; The phyloseq object will be built without phylogenetic tree: 'phylo_tree' not provided.</span></span>
<span><span class="co">#&gt; 1 samples removed due to zero total read count or &lt; 500 'min_reads'.</span></span>
<span><span class="co">#&gt; 28 samples found in common between 28 rows in 'X' and 29 rows in 'sample_data'.</span></span>
<span><span class="co">#&gt; 372 out of 2237 features were kept after 10 % prevalence filter.</span></span>
<span><span class="co">#&gt; Zeros will be imputed with 'zCompositions::cmultRepl()' prior to clr-transformation.</span></span></code></pre></div>
<p>When <code>verbose = TRUE</code> we can see that some useful
information is printed in the console (see <code><a href="../reference/process_ngs.html">process_ngs()</a></code>
for all settings and parameters options). For example, that a
phylogenetic tree has not been provided, that 1 sample has been removed
due to falling under the default 500 reads threshold for sequencing
depth, and that 372 features were kept after filtering for 10 %
prevalence.</p>
<p><code><a href="../reference/process_ngs.html">process_ngs()</a></code> generates a list object that contains:</p>
<ul>
<li>The processed matrix <code>X_processed</code> (in this case is
clr-transformed).</li>
<li>The matched sample_data (<code>sdata_final</code>).</li>
<li>A <code>phyloseq_raw</code> object that contains phyloseq objects
for all taxonomic ranks (see <code><a href="../reference/build_phyloseq.html">build_phyloseq()</a></code> for more
information).</li>
<li>A <code>NULL</code> <code>phyloseq_eco</code> object, because counts
have not been adjusted by microbial load. Counts adjusted by microbial
load are no longer compositional and they are referred as “ecosystem
counts”, hence the “_eco” suffix.</li>
</ul>
<p>We will see how to apply the processed data for common omics
analysis.</p>
</div>
<div class="section level3">
<h3 id="pca-visualisation-with-mva">PCA visualisation with <code>mva()</code><a class="anchor" aria-label="anchor" href="#pca-visualisation-with-mva"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mva.html">mva</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">X_processed</span>, </span>
<span>           sample_data <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">sdata_final</span>, </span>
<span>           group_colour <span class="op">=</span> <span class="st">"groups_liver"</span>, </span>
<span>           plot_title <span class="op">=</span> <span class="st">"Beta diversity (Aitchison)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 28 samples found in common between 28 rows in 'X' and 28 rows in 'sample_data'.</span></span>
<span><span class="co">#&gt; PCA</span></span>
<span><span class="co">#&gt; 28 samples x 372 variables</span></span>
<span><span class="co">#&gt; standard scaling of predictors</span></span>
<span><span class="co">#&gt;       R2X(cum) pre ort</span></span>
<span><span class="co">#&gt; Total    0.533   9   0</span></span></code></pre></div>
<p><img src="readyomics_files/figure-html/unnamed-chunk-12-1.png" alt="Figure panel with 4 plots, principal components variance percentage (upper-left), observation diagnostics for outliers (upper-right), scores plot (lower-left), loadings plot (lower right)." width="700"></p>
<p>By default, <code><a href="../reference/mva.html">mva()</a></code> will show a PCA diagnostics plot,
which is very useful to examine potential outliers, batch effects and
clusters related to the groups of interest.</p>
<p>In addition to storing the <code><a href="https://rdrr.io/pkg/ropls/man/opls.html" class="external-link">ropls::opls()</a></code> object,
<code><a href="../reference/mva.html">mva()</a></code> generates a publication-ready scores plot, in this
case color-coded by the variable <code>groups_liver</code> as we
indicated:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca</span><span class="op">$</span><span class="va">scores_plot</span></span></code></pre></div>
<p><img src="readyomics_files/figure-html/unnamed-chunk-13-1.png" alt="Scores plot color-coded by groups_liver class." width="700"></p>
<p>Because the plot is generated using ggplot2, it can be further
tweaked using ggplot2 functions before export.</p>
</div>
<div class="section level3">
<h3 id="permutational-analysis-of-variance-with-permanova">Permutational analysis of variance with
<code>permanova()</code><a class="anchor" aria-label="anchor" href="#permutational-analysis-of-variance-with-permanova"></a>
</h3>
<p>PERMANOVA is quite popular in microbiome datasets to inspect
differences in beta diversity, but it can also be applied to any type of
omics data.</p>
<p>By default, the function sets parameter
<code>indenpendent = TRUE</code>, meaning <code><a href="../reference/permanova.html">permanova()</a></code> also
calculates the overall variance for each covariable in the model
independently, in addition to the <code><a href="https://vegandevs.github.io/vegan/reference/adonis.html" class="external-link">vegan::adonis2()</a></code>
output.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define model</span></span>
<span><span class="va">rhs_model</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">groups_liver</span> <span class="op">+</span> <span class="va">alcohol_imp</span> <span class="op">+</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">age</span> <span class="op">+</span> <span class="va">PPI</span> <span class="op">+</span> <span class="va">smoking</span></span>
<span></span>
<span><span class="co"># Run PERMANOVA</span></span>
<span><span class="va">pova</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/permanova.html">permanova</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">X_processed</span>, </span>
<span>                  sample_data <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">sdata_final</span>,</span>
<span>                  formula_rhs <span class="op">=</span> <span class="va">rhs_model</span>, </span>
<span>                  platform <span class="op">=</span> <span class="st">"ngs"</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"Taxonomy"</span>, </span>
<span>                  seed <span class="op">=</span> <span class="fl">165</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in permanova(X = asv_ready$X_processed, sample_data = asv_ready$sdata_final, : 'independent = TRUE' but no variable-specific permutation controls found. Default 'perm_control$joint_terms' will be used.</span></span>
<span><span class="co">#&gt; 28 samples found in common between 28 rows in 'X' and 28 rows in 'sample_data'.</span></span>
<span><span class="co">#&gt; Permutation parameters for groups_liver have not been specified. Default 'perm_control$joint_terms' will be used.</span></span>
<span><span class="co">#&gt; Permutation parameters for alcohol_imp have not been specified. Default 'perm_control$joint_terms' will be used.</span></span>
<span><span class="co">#&gt; Permutation parameters for sex have not been specified. Default 'perm_control$joint_terms' will be used.</span></span>
<span><span class="co">#&gt; Permutation parameters for age have not been specified. Default 'perm_control$joint_terms' will be used.</span></span>
<span><span class="co">#&gt; Permutation parameters for PPI have not been specified. Default 'perm_control$joint_terms' will be used.</span></span>
<span><span class="co">#&gt; Permutation parameters for smoking have not been specified. Default 'perm_control$joint_terms' will be used.</span></span></code></pre></div>
<p>We see a warning and related verbose message (when
<code>verbose = TRUE</code>; the default) about permutation parameters
being used for all the variables. This is because you can set up
specific permutation control parameters for each covariable with
<code>perm_control</code>. This is particularly useful for example in
longitudinal studies, where you might need to apply permutation blocks
according to certain variables. In our case, we are good with the
default 999 unrestricted random permutations, so we can ignore the
warning and verbose. See <code><a href="../reference/permanova.html">permanova()</a></code> for more
information.</p>
<p>The output is a list that contains the distance matrix and
permutation matrix used, as well as two tables of results
<code>permanova_joint</code>, which is the default
<code><a href="https://vegandevs.github.io/vegan/reference/adonis.html" class="external-link">vegan::adonis2()</a></code> result, and <code>permanova_indep</code>,
when <code>independent = TRUE</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pova</span><span class="op">$</span><span class="va">permanova_joint</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table class="table">
<colgroup>
<col width="13%">
<col width="4%">
<col width="15%">
<col width="13%">
<col width="13%">
<col width="10%">
<col width="13%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">Df</th>
<th align="right">SumOfSqs</th>
<th align="right">R2</th>
<th align="right">F</th>
<th align="right">Pr(&gt;F)</th>
<th align="left">platform</th>
<th align="left">assay</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Model</td>
<td align="right">8</td>
<td align="right">4642.217</td>
<td align="right">0.313521</td>
<td align="right">1.084683</td>
<td align="right">0.09</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">Residual</td>
<td align="right">19</td>
<td align="right">10164.501</td>
<td align="right">0.686479</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">Total</td>
<td align="right">27</td>
<td align="right">14806.718</td>
<td align="right">1.000000</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pova</span><span class="op">$</span><span class="va">permanova_indep</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table class="table">
<colgroup>
<col width="23%">
<col width="3%">
<col width="11%">
<col width="10%">
<col width="10%">
<col width="7%">
<col width="13%">
<col width="9%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">Df</th>
<th align="right">SumOfSqs</th>
<th align="right">R2</th>
<th align="right">F</th>
<th align="right">Pr(&gt;F)</th>
<th align="left">var_name</th>
<th align="left">platform</th>
<th align="left">assay</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">groups_liver.Model</td>
<td align="right">1</td>
<td align="right">461.8705</td>
<td align="right">0.0311933</td>
<td align="right">0.8371392</td>
<td align="right">0.940</td>
<td align="left">groups_liver</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">groups_liver.Residual</td>
<td align="right">26</td>
<td align="right">14344.8473</td>
<td align="right">0.9688067</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">groups_liver.Total</td>
<td align="right">27</td>
<td align="right">14806.7179</td>
<td align="right">1.0000000</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">alcohol_imp.Model</td>
<td align="right">2</td>
<td align="right">1088.7284</td>
<td align="right">0.0735294</td>
<td align="right">0.9920627</td>
<td align="right">0.499</td>
<td align="left">alcohol_imp</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">alcohol_imp.Residual</td>
<td align="right">25</td>
<td align="right">13717.9895</td>
<td align="right">0.9264706</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">alcohol_imp</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">alcohol_imp.Total</td>
<td align="right">27</td>
<td align="right">14806.7179</td>
<td align="right">1.0000000</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">alcohol_imp</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">sex.Model</td>
<td align="right">1</td>
<td align="right">545.4117</td>
<td align="right">0.0368354</td>
<td align="right">0.9943482</td>
<td align="right">0.439</td>
<td align="left">sex</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">sex.Residual</td>
<td align="right">26</td>
<td align="right">14261.3062</td>
<td align="right">0.9631646</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">sex</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">sex.Total</td>
<td align="right">27</td>
<td align="right">14806.7179</td>
<td align="right">1.0000000</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">sex</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">age.Model</td>
<td align="right">1</td>
<td align="right">527.1507</td>
<td align="right">0.0356021</td>
<td align="right">0.9598272</td>
<td align="right">0.580</td>
<td align="left">age</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">age.Residual</td>
<td align="right">26</td>
<td align="right">14279.5672</td>
<td align="right">0.9643979</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">age</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">age.Total</td>
<td align="right">27</td>
<td align="right">14806.7179</td>
<td align="right">1.0000000</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">age</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">PPI.Model</td>
<td align="right">1</td>
<td align="right">653.3245</td>
<td align="right">0.0441235</td>
<td align="right">1.2001672</td>
<td align="right">0.098</td>
<td align="left">PPI</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">PPI.Residual</td>
<td align="right">26</td>
<td align="right">14153.3934</td>
<td align="right">0.9558765</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">PPI</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">PPI.Total</td>
<td align="right">27</td>
<td align="right">14806.7179</td>
<td align="right">1.0000000</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">PPI</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">smoking.Model</td>
<td align="right">2</td>
<td align="right">1421.2488</td>
<td align="right">0.0959868</td>
<td align="right">1.3272310</td>
<td align="right">0.007</td>
<td align="left">smoking</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="odd">
<td align="left">smoking.Residual</td>
<td align="right">25</td>
<td align="right">13385.4691</td>
<td align="right">0.9040132</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">smoking</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
<tr class="even">
<td align="left">smoking.Total</td>
<td align="right">27</td>
<td align="right">14806.7179</td>
<td align="right">1.0000000</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">smoking</td>
<td align="left">ngs</td>
<td align="left">Taxonomy</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="differential-analysis-with-dana">
<u>D</u>ifferential <u>ANA</u>lysis with <code>dana()</code><a class="anchor" aria-label="anchor" href="#differential-analysis-with-dana"></a>
</h3>
<p>A common analysis goal with omics data is to try and find omic
features associated with traits of interest. <code><a href="../reference/dana.html">dana()</a></code> is
readyomics function for differential analysis. At present,
<code><a href="../reference/dana.html">dana()</a></code> supports linear regressions with
<code><a href="https://rdrr.io/r/stats/lm.html" class="external-link">stats::lm()</a></code> and linear mixed effects models with
<code><a href="https://rdrr.io/pkg/lme4/man/lmer.html" class="external-link">lme4::lmer()</a></code>. <code><a href="../reference/dana.html">dana()</a></code> main strengths are:</p>
<ul>
<li><p>Full flexibility on model formula (random effects, interaction
terms).</p></li>
<li>
<p>Likelihood ratio tests (LRT) in addition to coefficient P
values.</p>
<ul>
<li>LRT is especially recommended for small studies (approx. below 100
samples).</li>
<li>Any formula terms can be tested with LRT (interaction terms, random
effects).</li>
<li>Several terms can be tested at once
(e.g. <code>term_LRT = c("sex" , "BMI", "1 | country")</code>).</li>
</ul>
</li>
<li><p>Easy set up to run in parallel using <code><a href="https://future.futureverse.org/reference/plan.html" class="external-link">future::plan()</a></code>
- user has full control, allowing good scalability for big
data.</p></li>
<li><p>Toggable progress bar with progressr.</p></li>
</ul>
<p>We will fit a linear regression with the main variables in our
dataset:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># future::plan(multisession, workers = 4) # For running 4 processes in parallel</span></span>
<span><span class="co"># progressr::handlers(global = TRUE) # To show progress bar</span></span>
<span><span class="va">dana_asv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dana.html">dana</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">X_processed</span>, </span>
<span>                 sample_data <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">sdata_final</span>,</span>
<span>                 formula_rhs <span class="op">=</span> <span class="va">rhs_model</span>, </span>
<span>                 term_LRT <span class="op">=</span> <span class="st">"groups_liver"</span>,</span>
<span>                 platform <span class="op">=</span> <span class="st">"ngs"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in dana(X = asv_ready$X_processed, sample_data = asv_ready$sdata_final, : Set 'future::plan' parallel strategy to speed up computation.</span></span>
<span><span class="co">#&gt; 28 samples found in common between 28 rows in 'X' and 28 rows in 'sample_data'.</span></span>
<span><span class="co">#&gt; Using model: stats::lm.</span></span>
<span><span class="co"># plan(sequential) # To disable parallel processing</span></span></code></pre></div>
<p>We first see a warning message suggesting to use parallel computing
due to the <code>X</code> matrix having more than 100 features. However
this can be ignored as <code><a href="https://rdrr.io/r/stats/lm.html" class="external-link">stats::lm()</a></code> is very quick
anyway.</p>
<p>The <code><a href="../reference/dana.html">dana()</a></code> function generates a <code>dana</code> object
including the input data used, and the publication-ready
<code>fit</code> table summarizing the main model results for each
feature analysed (<code>"feat_id"</code> column).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dana_asv</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table class="table">
<colgroup>
<col width="18%">
<col width="10%">
<col width="10%">
<col width="10%">
<col width="17%">
<col width="10%">
<col width="9%">
<col width="7%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">Coefficient</th>
<th align="right">Estimate</th>
<th align="right">Std. Error</th>
<th align="right">t value</th>
<th align="right">Pr(&gt;|t|)</th>
<th align="right">2.5 %</th>
<th align="right">97.5 %</th>
<th align="left">feat_id</th>
<th align="left">platform</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">5.7417549</td>
<td align="right">0.9926354</td>
<td align="right">5.7843541</td>
<td align="right">0.0000142</td>
<td align="right">3.6641450</td>
<td align="right">7.8193648</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
<tr class="even">
<td align="left">groups_liverFL_HS</td>
<td align="right">0.2230096</td>
<td align="right">0.2429450</td>
<td align="right">0.9179426</td>
<td align="right">0.3701552</td>
<td align="right">-0.2854802</td>
<td align="right">0.7314994</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
<tr class="odd">
<td align="left">alcohol_impexcess</td>
<td align="right">-0.0880760</td>
<td align="right">0.4729165</td>
<td align="right">-0.1862400</td>
<td align="right">0.8542306</td>
<td align="right">-1.0779017</td>
<td align="right">0.9017497</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
<tr class="even">
<td align="left">alcohol_impmoderate</td>
<td align="right">-0.1708858</td>
<td align="right">0.4263501</td>
<td align="right">-0.4008109</td>
<td align="right">0.6930308</td>
<td align="right">-1.0632467</td>
<td align="right">0.7214752</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
<tr class="odd">
<td align="left">sex2</td>
<td align="right">-0.4434339</td>
<td align="right">0.2604760</td>
<td align="right">-1.7023984</td>
<td align="right">0.1049857</td>
<td align="right">-0.9886163</td>
<td align="right">0.1017486</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
<tr class="even">
<td align="left">age</td>
<td align="right">-0.0102866</td>
<td align="right">0.0147685</td>
<td align="right">-0.6965206</td>
<td align="right">0.4945367</td>
<td align="right">-0.0411974</td>
<td align="right">0.0206243</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
</tbody>
</table>
</div>
<p>If working with multiple omics, is useful to also specify
<code>platform</code> and <code>assay</code> columns for seamless
downstream omics results integration.</p>
<p>When one or more terms are specified for likelihood ratio testing
with <code>term_LRT</code>, <code>dana</code> also includes the
<code>lrt</code> table with the LRT results. In our case we used LRT to
test <code>groups_liver</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dana_asv</span><span class="op">$</span><span class="va">lrt</span><span class="op">)</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table class="table">
<colgroup>
<col width="8%">
<col width="12%">
<col width="3%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="16%">
<col width="10%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="right">Res.Df</th>
<th align="right">RSS</th>
<th align="right">Df</th>
<th align="right">Sum of Sq</th>
<th align="right">F</th>
<th align="right">Pr(&gt;F)</th>
<th align="left">term</th>
<th align="left">feat_id</th>
<th align="left">platform</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">20</td>
<td align="right">6.641948</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
<tr class="even">
<td align="right">19</td>
<td align="right">6.359897</td>
<td align="right">1</td>
<td align="right">0.2820509</td>
<td align="right">0.8426186</td>
<td align="right">0.3701552</td>
<td align="left">groups_liver</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
</tr>
<tr class="odd">
<td align="right">20</td>
<td align="right">70.044952</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ASV2</td>
<td align="left">ngs</td>
</tr>
<tr class="even">
<td align="right">19</td>
<td align="right">69.880252</td>
<td align="right">1</td>
<td align="right">0.1647003</td>
<td align="right">0.0447810</td>
<td align="right">0.8346607</td>
<td align="left">groups_liver</td>
<td align="left">ASV2</td>
<td align="left">ngs</td>
</tr>
<tr class="odd">
<td align="right">20</td>
<td align="right">8.220301</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ASV3</td>
<td align="left">ngs</td>
</tr>
<tr class="even">
<td align="right">19</td>
<td align="right">8.044147</td>
<td align="right">1</td>
<td align="right">0.1761536</td>
<td align="right">0.4160687</td>
<td align="right">0.5266180</td>
<td align="left">groups_liver</td>
<td align="left">ASV3</td>
<td align="left">ngs</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="adjusting-for-multiple-comparisons-with-adjust_pval">Adjusting for multiple comparisons with
<code>adjust_pval()</code><a class="anchor" aria-label="anchor" href="#adjusting-for-multiple-comparisons-with-adjust_pval"></a>
</h4>
<p>We leverage R base pipes to conveniently add adjusted P values to the
<code>dana</code> object results with <code><a href="../reference/adjust_pval.html">adjust_pval()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dana_asv</span> <span class="op">&lt;-</span> <span class="va">dana_asv</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/adjust_pval.html">adjust_pval</a></span><span class="op">(</span>padj_by <span class="op">=</span> <span class="st">"terms"</span>, </span>
<span>              padj_method <span class="op">=</span> <span class="st">"BH"</span>, </span>
<span>              padj_method_LRT <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>fit</code> table now has new columns with the adjusted P
values. Because we chose to adjust by terms in the <code>fit</code>
table (<code>padj_by = "terms"</code>), this will adjust each covariable
P value individually. We see more than one column of adjusted P values
with the format
<code>padj_[</code><em><code>padj_method</code></em><code>]_[</code><em><code>term</code></em><code>]</code>
(e.g. <code>"padj_BH_sex2"</code>).</p>
<p>The alternative would be (<code>padj_by = "all"</code>) in which all
nominal P values in <code>fit</code> will be adjusted together. In this
case, a single column would be added to <code>fit</code>, with the
format
<code>padj_[</code><em><code>padj_method</code></em><code>]</code>.</p>
<p>If a formula fixed term was also tested via LRT,
<code><a href="../reference/adjust_pval.html">adjust_pval()</a></code> will also add the adjusted LRT P value to the
<code>fit</code> table. LRT P values have the <code>"_LRT"</code> suffix
(see last column <code>"padj_BH_groups_liver_LRT"</code>).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dana_asv</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table style="width:100%;" class="table">
<colgroup>
<col width="6%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="6%">
<col width="3%">
<col width="3%">
<col width="2%">
<col width="3%">
<col width="9%">
<col width="9%">
<col width="9%">
<col width="4%">
<col width="4%">
<col width="4%">
<col width="5%">
<col width="5%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">Coefficient</th>
<th align="right">Estimate</th>
<th align="right">Std. Error</th>
<th align="right">t value</th>
<th align="right">Pr(&gt;|t|)</th>
<th align="right">2.5 %</th>
<th align="right">97.5 %</th>
<th align="left">feat_id</th>
<th align="left">platform</th>
<th align="right">padj_BH_groups_liverFL_HS</th>
<th align="right">padj_BH_alcohol_impexcess</th>
<th align="right">padj_BH_alcohol_impmoderate</th>
<th align="right">padj_BH_sex2</th>
<th align="right">padj_BH_age</th>
<th align="right">padj_BH_PPI2</th>
<th align="right">padj_BH_smoking2</th>
<th align="right">padj_BH_smoking3</th>
<th align="right">padj_BH_groups_liver_LRT</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">5.7417549</td>
<td align="right">0.9926354</td>
<td align="right">5.7843541</td>
<td align="right">0.0000142</td>
<td align="right">3.6641450</td>
<td align="right">7.8193648</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">groups_liverFL_HS</td>
<td align="right">0.2230096</td>
<td align="right">0.2429450</td>
<td align="right">0.9179426</td>
<td align="right">0.3701552</td>
<td align="right">-0.2854802</td>
<td align="right">0.7314994</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">0.9252859</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.9252859</td>
</tr>
<tr class="odd">
<td align="left">alcohol_impexcess</td>
<td align="right">-0.0880760</td>
<td align="right">0.4729165</td>
<td align="right">-0.1862400</td>
<td align="right">0.8542306</td>
<td align="right">-1.0779017</td>
<td align="right">0.9017497</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">0.9986091</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">alcohol_impmoderate</td>
<td align="right">-0.1708858</td>
<td align="right">0.4263501</td>
<td align="right">-0.4008109</td>
<td align="right">0.6930308</td>
<td align="right">-1.0632467</td>
<td align="right">0.7214752</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.955298</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">sex2</td>
<td align="right">-0.4434339</td>
<td align="right">0.2604760</td>
<td align="right">-1.7023984</td>
<td align="right">0.1049857</td>
<td align="right">-0.9886163</td>
<td align="right">0.1017486</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.8523934</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">age</td>
<td align="right">-0.0102866</td>
<td align="right">0.0147685</td>
<td align="right">-0.6965206</td>
<td align="right">0.4945367</td>
<td align="right">-0.0411974</td>
<td align="right">0.0206243</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.9220974</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
</tbody>
</table>
</div>
<p>Similarly, we can inspect the <code>lrt</code> table with the added
adjusted P values.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dana_asv</span><span class="op">$</span><span class="va">lrt</span><span class="op">)</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table style="width:100%;" class="table">
<colgroup>
<col width="6%">
<col width="9%">
<col width="2%">
<col width="9%">
<col width="9%">
<col width="9%">
<col width="12%">
<col width="7%">
<col width="8%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th align="right">Res.Df</th>
<th align="right">RSS</th>
<th align="right">Df</th>
<th align="right">Sum of Sq</th>
<th align="right">F</th>
<th align="right">Pr(&gt;F)</th>
<th align="left">term</th>
<th align="left">feat_id</th>
<th align="left">platform</th>
<th align="right">padj_BH_groups_liver</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">20</td>
<td align="right">6.641948</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="right">19</td>
<td align="right">6.359897</td>
<td align="right">1</td>
<td align="right">0.2820509</td>
<td align="right">0.8426186</td>
<td align="right">0.3701552</td>
<td align="left">groups_liver</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">0.9252859</td>
</tr>
<tr class="odd">
<td align="right">20</td>
<td align="right">70.044952</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ASV2</td>
<td align="left">ngs</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="right">19</td>
<td align="right">69.880252</td>
<td align="right">1</td>
<td align="right">0.1647003</td>
<td align="right">0.0447810</td>
<td align="right">0.8346607</td>
<td align="left">groups_liver</td>
<td align="left">ASV2</td>
<td align="left">ngs</td>
<td align="right">0.9949226</td>
</tr>
<tr class="odd">
<td align="right">20</td>
<td align="right">8.220301</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">groups_liver</td>
<td align="left">ASV3</td>
<td align="left">ngs</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="right">19</td>
<td align="right">8.044147</td>
<td align="right">1</td>
<td align="right">0.1761536</td>
<td align="right">0.4160687</td>
<td align="right">0.5266180</td>
<td align="left">groups_liver</td>
<td align="left">ASV3</td>
<td align="left">ngs</td>
<td align="right">0.9379068</td>
</tr>
</tbody>
</table>
</div>
<p>While a single adjustment method should be chosen <em>a priori</em>,
it is possible to add more than one P value adjustment method
(e.g. <code>padj_method = c("BH", "storey")</code>,
<code>padj_method_LRT = c("BH", "bonferroni")</code>). See
<code><a href="../reference/adjust_pval.html">adjust_pval()</a></code> for more information.</p>
</div>
<div class="section level4">
<h4 id="adding-feature-labels-with-add_feat_name-or-add_taxa">Adding feature labels with <code>add_feat_name()</code> or
<code>add_taxa()</code><a class="anchor" aria-label="anchor" href="#adding-feature-labels-with-add_feat_name-or-add_taxa"></a>
</h4>
<p>Original labels for some omics data can contain non-alphanumeric
characters and symbols that could be problematic to fit
<code><a href="../reference/dana.html">dana()</a></code> models. <code><a href="../reference/process_ms.html">process_ms()</a></code> allows to rename
features as <code>"feat_1"</code>, <code>"feat_2"</code>, and so on, by
setting <code>rename_feat = TRUE</code>, while storing the original
labels.</p>
<p><code><a href="../reference/add_feat_name.html">add_feat_name()</a></code> can then be used to add the original
labels to the <code>dana</code> object before plotting significant
results.</p>
<p>Similarly, for metataxonomics and metagenomics data, it is more
interesting to read the genus/species/strain of origin rather than a
cryptic <code>"ASV1"</code>. In this case, we would use
<code><a href="../reference/add_taxa.html">add_taxa()</a></code>. Which also adds the corresponding higher
hierarchy taxa names.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dana_asv</span> <span class="op">&lt;-</span> <span class="va">dana_asv</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/add_taxa.html">add_taxa</a></span><span class="op">(</span>taxa_table <span class="op">=</span> <span class="va">taxa</span>, </span>
<span>           taxa_rank <span class="op">=</span> <span class="st">"asv"</span><span class="op">)</span></span></code></pre></div>
<p>We can see that various columns have been added to the
<code>fit</code> table. Because <code>taxa_rank = "asv"</code>, the
column for the feature label <code>taxon_name</code> by default adds the
species (if provided) or the genus name to the ASV ID, collapsed by “_“,
a format quite common in publication plots.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dana_asv</span><span class="op">$</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<div style="overflow-x: auto;">
<table style="width:100%;" class="table">
<colgroup>
<col width="5%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="4%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="6%">
<col width="6%">
<col width="7%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="6%">
<col width="2%">
<col width="3%">
<col width="2%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th align="left">Coefficient</th>
<th align="right">Estimate</th>
<th align="right">Std. Error</th>
<th align="right">t value</th>
<th align="right">Pr(&gt;|t|)</th>
<th align="right">2.5 %</th>
<th align="right">97.5 %</th>
<th align="left">feat_id</th>
<th align="left">platform</th>
<th align="right">padj_BH_groups_liverFL_HS</th>
<th align="right">padj_BH_alcohol_impexcess</th>
<th align="right">padj_BH_alcohol_impmoderate</th>
<th align="right">padj_BH_sex2</th>
<th align="right">padj_BH_age</th>
<th align="right">padj_BH_PPI2</th>
<th align="right">padj_BH_smoking2</th>
<th align="right">padj_BH_smoking3</th>
<th align="right">padj_BH_groups_liver_LRT</th>
<th align="left">taxon_rank</th>
<th align="left">taxon_name</th>
<th align="left">species</th>
<th align="left">genus</th>
<th align="left">family</th>
<th align="left">order</th>
<th align="left">class</th>
<th align="left">phylum</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">5.7417549</td>
<td align="right">0.9926354</td>
<td align="right">5.7843541</td>
<td align="right">0.0000142</td>
<td align="right">3.6641450</td>
<td align="right">7.8193648</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">asv</td>
<td align="left">vulgatus_ASV1</td>
<td align="left">vulgatus</td>
<td align="left">Bacteroides</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidota</td>
</tr>
<tr class="even">
<td align="left">groups_liverFL_HS</td>
<td align="right">0.2230096</td>
<td align="right">0.2429450</td>
<td align="right">0.9179426</td>
<td align="right">0.3701552</td>
<td align="right">-0.2854802</td>
<td align="right">0.7314994</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">0.9252859</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.9252859</td>
<td align="left">asv</td>
<td align="left">vulgatus_ASV1</td>
<td align="left">vulgatus</td>
<td align="left">Bacteroides</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidota</td>
</tr>
<tr class="odd">
<td align="left">alcohol_impexcess</td>
<td align="right">-0.0880760</td>
<td align="right">0.4729165</td>
<td align="right">-0.1862400</td>
<td align="right">0.8542306</td>
<td align="right">-1.0779017</td>
<td align="right">0.9017497</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">0.9986091</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">asv</td>
<td align="left">vulgatus_ASV1</td>
<td align="left">vulgatus</td>
<td align="left">Bacteroides</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidota</td>
</tr>
<tr class="even">
<td align="left">alcohol_impmoderate</td>
<td align="right">-0.1708858</td>
<td align="right">0.4263501</td>
<td align="right">-0.4008109</td>
<td align="right">0.6930308</td>
<td align="right">-1.0632467</td>
<td align="right">0.7214752</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.955298</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">asv</td>
<td align="left">vulgatus_ASV1</td>
<td align="left">vulgatus</td>
<td align="left">Bacteroides</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidota</td>
</tr>
<tr class="odd">
<td align="left">sex2</td>
<td align="right">-0.4434339</td>
<td align="right">0.2604760</td>
<td align="right">-1.7023984</td>
<td align="right">0.1049857</td>
<td align="right">-0.9886163</td>
<td align="right">0.1017486</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.8523934</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">asv</td>
<td align="left">vulgatus_ASV1</td>
<td align="left">vulgatus</td>
<td align="left">Bacteroides</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidota</td>
</tr>
<tr class="even">
<td align="left">age</td>
<td align="right">-0.0102866</td>
<td align="right">0.0147685</td>
<td align="right">-0.6965206</td>
<td align="right">0.4945367</td>
<td align="right">-0.0411974</td>
<td align="right">0.0206243</td>
<td align="left">ASV1</td>
<td align="left">ngs</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.9220974</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">asv</td>
<td align="left">vulgatus_ASV1</td>
<td align="left">vulgatus</td>
<td align="left">Bacteroides</td>
<td align="left">Bacteroidaceae</td>
<td align="left">Bacteroidales</td>
<td align="left">Bacteroidia</td>
<td align="left">Bacteroidota</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level4">
<h4 id="visualise-results-with-ready_plots">Visualise results with <code>ready_plots()</code><a class="anchor" aria-label="anchor" href="#visualise-results-with-ready_plots"></a>
</h4>
<p>We can pipe <code><a href="../reference/ready_plots.html">ready_plots()</a></code> to the previous command, or
skip the previous step and directly apply <code><a href="../reference/ready_plots.html">ready_plots()</a></code> to
the <code>dana</code> object to explore the results with the default
feature IDs (<code>"feat_id"</code> column).</p>
<p><code><a href="../reference/ready_plots.html">ready_plots()</a></code> automatically generates and stores the
most commonly used publication-ready plots:</p>
<ul>
<li>Differential analysis significant coefficients (volcano, dotplot,
heatmap - capped at 50 features),</li>
<li>Significant features abundance (capped at the 10 most significant,
but plots for specific features can be requested with
<code>X_colnames</code> parameter).</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generates error due to lack of significant features:</span></span>
<span><span class="va">dana_asv</span> <span class="op">&lt;-</span> <span class="va">dana_asv</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/ready_plots.html">ready_plots</a></span><span class="op">(</span>term_name <span class="op">=</span> <span class="st">"groups_liver"</span>, <span class="co"># Formula fit term of interest</span></span>
<span>              pval_match <span class="op">=</span> <span class="st">"groups_liver_LRT"</span>, <span class="co"># LRT adjusted P values will be used</span></span>
<span>              sdata_var <span class="op">=</span> <span class="st">"groups_liver"</span>, <span class="co"># Grouping variable for individual feature plots</span></span>
<span>              group_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FL <span class="op">=</span> <span class="st">"#4daf4a"</span>, <span class="co"># Optional color customization</span></span>
<span>                                FL_HS <span class="op">=</span> <span class="st">"#377eb8"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in ready_plots(dana_asv, term_name = "groups_liver", pval_match = "groups_liver_LRT", : No significant results at selected 0.1 significance threshold.</span></span></code></pre></div>
<p>In our example, no features were significant after multiple
comparison correction. This prompts an error message from
<code><a href="../reference/ready_plots.html">ready_plots()</a></code> which halts execution. This is somewhat
expected as we are using a small portion of the original dataset as a
toy example.</p>
<p>A formal analysis would end here, concluding that the null hypothesis
could not be rejected. In our case, we will use the nominal P value to
showcase the <code><a href="../reference/ready_plots.html">ready_plots()</a></code> functionality.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dana_asv</span> <span class="op">&lt;-</span> <span class="va">dana_asv</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/ready_plots.html">ready_plots</a></span><span class="op">(</span>term_name <span class="op">=</span> <span class="st">"groups_liver"</span>,</span>
<span>              pval_match <span class="op">=</span> <span class="st">"Pr"</span>, <span class="co"># Nominal P value for illustration purposes</span></span>
<span>              sdata_var <span class="op">=</span> <span class="st">"groups_liver"</span>, </span>
<span>              group_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FL <span class="op">=</span> <span class="st">"#4daf4a"</span>, </span>
<span>                                FL_HS <span class="op">=</span> <span class="st">"#377eb8"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 33 significant results found at selected 0.1 significance threshold.</span></span>
<span><span class="co">#&gt; 'X_colnames' = NULL: the first 10 most significant features will be plotted.</span></span></code></pre></div>
<p>When <code>verbose = TRUE</code> (the default)
<code><a href="../reference/ready_plots.html">ready_plots()</a></code> displays the total number of significant
features at the chosen P value threshold. We can inspect all plots at
once by simply running:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dana_asv</span><span class="op">$</span><span class="va">plots</span></span>
<span><span class="co">#&gt; $coeff_volcano</span></span>
<span><span class="co">#&gt; Warning: Removed 339 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_text_repel()`).</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 7 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="readyomics_files/figure-html/unnamed-chunk-34-1.png" alt="Plots for model coefficients (volcano plot, heatmap, points) and plots for  significant features across study groups (boxplot, violin and ridges plots)." width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $coeff_heatmap</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-34-2.png" alt="Plots for model coefficients (volcano plot, heatmap, points) and plots for  significant features across study groups (boxplot, violin and ridges plots)." width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $coeff_point</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-34-3.png" alt="Plots for model coefficients (volcano plot, heatmap, points) and plots for  significant features across study groups (boxplot, violin and ridges plots)." width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $feat_boxplot</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-34-4.png" alt="Plots for model coefficients (volcano plot, heatmap, points) and plots for  significant features across study groups (boxplot, violin and ridges plots)." width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $feat_violin</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-34-5.png" alt="Plots for model coefficients (volcano plot, heatmap, points) and plots for  significant features across study groups (boxplot, violin and ridges plots)." width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $feat_ridge</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.471</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-34-6.png" alt="Plots for model coefficients (volcano plot, heatmap, points) and plots for  significant features across study groups (boxplot, violin and ridges plots)." width="700"></p>
<p>Because plots are generated with ggplot2, they can easily be further
customized.</p>
</div>
</div>
<div class="section level3">
<h3 id="bonus-microbiome-dana-for-alpha-diversity">Bonus (microbiome): <code>dana()</code> for alpha diversity<a class="anchor" aria-label="anchor" href="#bonus-microbiome-dana-for-alpha-diversity"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute alpha diversity measures for the ASV phyloseq object</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/estimate_richness.html" class="external-link">estimate_richness</a></span><span class="op">(</span><span class="va">asv_ready</span><span class="op">$</span><span class="va">phyloseq_raw</span><span class="op">$</span><span class="va">asv</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu">matches</span><span class="op">(</span><span class="st">"ace|chao|fisher"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate alpha diversity differences among groups with dana()</span></span>
<span><span class="va">dana_alpha</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dana.html">dana</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">alpha</span>, </span>
<span>                   sample_data <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">sdata_final</span>,</span>
<span>                   formula_rhs <span class="op">=</span> <span class="va">rhs_model</span>, </span>
<span>                   term_LRT <span class="op">=</span> <span class="st">"groups_liver"</span>,</span>
<span>                   platform <span class="op">=</span> <span class="st">"ngs"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 28 samples found in common between 28 rows in 'X' and 28 rows in 'sample_data'.</span></span>
<span><span class="co">#&gt; Using model: stats::lm.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="bonus-microbiome-dana-for-higher-taxonomic-ranks">Bonus (microbiome): <code>dana()</code> for higher taxonomic
ranks<a class="anchor" aria-label="anchor" href="#bonus-microbiome-dana-for-higher-taxonomic-ranks"></a>
</h3>
<p>The following example shows only the code for genus rank, but this
code could be easily wrapped into a function that iterates across all
taxa ranks to analyse all at once, using parallel computation with
<code><a href="https://future.futureverse.org/reference/plan.html" class="external-link">future::plan()</a></code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># CLR-transform genus rank counts table</span></span>
<span><span class="va">genus_ready</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_ngs.html">process_ngs</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">asv_ready</span><span class="op">$</span><span class="va">phyloseq_raw</span><span class="op">$</span><span class="va">genus</span><span class="op">@</span><span class="va">otu_table</span><span class="op">)</span>, </span>
<span>                           sample_data <span class="op">=</span> <span class="va">asv_ready</span><span class="op">$</span><span class="va">sdata_final</span>, </span>
<span>                           taxa_table <span class="op">=</span> <span class="va">taxa</span>, </span>
<span>                           normalise <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                           transform <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                           raw_phyloseq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                           eco_phyloseq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run dana</span></span>
<span><span class="va">dana_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dana.html">dana</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">genus_ready</span><span class="op">$</span><span class="va">X_processed</span>, </span>
<span>                   sample_data <span class="op">=</span> <span class="va">genus_ready</span><span class="op">$</span><span class="va">sdata_final</span>,</span>
<span>                   formula_rhs <span class="op">=</span> <span class="va">rhs_model</span>, </span>
<span>                   term_LRT <span class="op">=</span> <span class="st">"groups_liver"</span>,</span>
<span>                   platform <span class="op">=</span> <span class="st">"ngs"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/adjust_pval.html">adjust_pval</a></span><span class="op">(</span>padj_by <span class="op">=</span> <span class="st">"terms"</span>, </span>
<span>              padj_method <span class="op">=</span> <span class="st">"BH"</span>, </span>
<span>              padj_method_LRT <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/add_taxa.html">add_taxa</a></span><span class="op">(</span>taxa_table <span class="op">=</span> <span class="va">taxa</span>, </span>
<span>           taxa_rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/ready_plots.html">ready_plots</a></span><span class="op">(</span>term_name <span class="op">=</span> <span class="st">"groups_liver"</span>,</span>
<span>              pval_match <span class="op">=</span> <span class="st">"Pr"</span>, <span class="co"># Nominal P value for illustration purposes</span></span>
<span>              sdata_var <span class="op">=</span> <span class="st">"groups_liver"</span>,</span>
<span>              group_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FL <span class="op">=</span> <span class="st">"#4daf4a"</span>,</span>
<span>                                FL_HS <span class="op">=</span> <span class="st">"#377eb8"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in dana(X = genus_ready$X_processed, sample_data = genus_ready$sdata_final, : Set 'future::plan' parallel strategy to speed up computation.</span></span>
<span><span class="co">#&gt; 28 samples found in common between 28 rows in 'X' and 28 rows in 'sample_data'.</span></span>
<span><span class="co">#&gt; Using model: stats::lm.</span></span>
<span><span class="co">#&gt; 9 significant results found at selected 0.1 significance threshold.</span></span>
<span><span class="co">#&gt; 'X_colnames' = NULL: the first 10 most significant features will be plotted.</span></span>
<span></span>
<span><span class="co"># Inspect plots</span></span>
<span><span class="va">dana_genus</span><span class="op">$</span><span class="va">plots</span></span>
<span><span class="co">#&gt; $coeff_volcano</span></span>
<span><span class="co">#&gt; Warning: Removed 97 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_text_repel()`).</span></span></code></pre></div>
<p><img src="readyomics_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $coeff_heatmap</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-36-2.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $coeff_point</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-36-3.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $feat_boxplot</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-36-4.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $feat_violin</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-36-5.png" width="700"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $feat_ridge</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.595</span></span></code></pre>
<p><img src="readyomics_files/figure-html/unnamed-chunk-36-6.png" width="700"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Laura Martinez-Gili.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
